### ISSUE TITLE
The same uint type constants and variables are converted to different floats

### ISSUE THREAD

**Issue by ffcr** [KhronosGroup/WebGL#3101]
In the standard OpenGL, when a==- 1u, whether a is a constant or a variable, float (a) is 4294967295.0, but in the glsl of webgl2, when a==- 1u, if a is a variable, float (a) will return to zero,if it is a constant, the result is the same as that of the standard OpenGL
Look here https://www.shadertoy.com/view/Ms3fz7

**Comment by aleino-nv** [comment]
I believe shadertoy uses WebGL 2.
Check the WebGL 2 specification [1], and you'll find in section "4.3 GLSL ES 3.00 support" that GLSL ES 3.0 is supported.
Seemingly, shadertoy somehow declares your shader as GLSL ES 3.0 because you're able to use the uint keyword. (It seems they paste in "#version es 300" at the top of your shader.)

Now check section 5.4.1 "Conversion and Scalar Constructors" in the GLSL ES 3.0 specification [2]:
"
When constructors are used to convert a float to an int or uint, the fractional part of the floating-point
value is dropped. It is undefined to convert a negative floating point value to an uint
"

In other words, the result is not defined so I don't think you're seeing a bug here, just undefined behavior.

[1] https://www.khronos.org/registry/webgl/specs/latest/2.0/
[2] https://www.khronos.org/registry/OpenGL/specs/es/3.0/GLSL_ES_Specification_3.00.pdf

**Comment by jdashg** [comment]
I agree, and I don't think we need to tighten this down for WebGL.
@kenrussell WDYT?

**Comment by kenrussell** [comment]
Agreed @jdashg - it's not feasible for WebGL to eliminate all of GLSL ES's undefined behavior. Doing so typically requires emulation code to be added to the shader, which will negatively impact performance.

@aleino-nv thank you for tracking down the root cause of this behavioral difference.
